### ISSUE TITLE
Path.rglob performance issues in deeply nested directories compared to glob.glob(recursive=True)

### ISSUE THREAD

**Issue by ionite34** [python/cpython#102613]
# Bug report

Pathlib.rglob can be orders of magnitudes slower than `glob.glob(recursive=True)`

With a 1000-deep nested directory, `glob.glob` and `Path.glob` both took under 1 second. `Path.rglob` took close to 1.5 minutes.

```python
import glob
import os
from pathlib import Path

x = ""
for _ in range(1000):
    x += "a/"
    os.mkdir(x)
    
# ~ 0.5s
print(glob.glob("**/*", recursive=True))

# ~ 87s
print(list(Path(".").rglob("**/*")))
```

### Linked PRs
* gh-104244
* gh-104373
* gh-104512
* gh-105744
* gh-105749

**Comment by pochmann** [comment]
Your code there isn't using `rglob`, is it?

Interesting, I guessed quadratic time from recursive yielding but it seems even worse than that. Depths and times:
```
100 0.21748781204223633
200 1.3257858753204346
400 7.545353412628174
100 0.23529744148254395
200 1.1740493774414062
400 7.466708183288574
100 0.23092246055603027
200 1.1659789085388184
400 7.2503981590271
```
<details><summary>Code</summary>

```python
import os
from pathlib import Path
from time import time
import shutil

for depth in [100, 200, 400] * 3:
    x = ""
    for _ in range(depth):
        x += "a/"
        os.mkdir(x)
    t = time()
    list(Path(".").rglob("**/*"))
    print(depth, time() - t)
    shutil.rmtree('a')
```

[Attempt This Online!](https://ato.pxeger.com/run?1=PVAxjsIwEBStn0C1coMdIOSAAiHxB3qEUBBOYl0cW5tFgrfQ0HCPutcQZwMurNnZmfV6Hn_hTpVvns_XlYr55n80ti54JPCtKNA7CDlVtT3DQO-7khtknfmwEYsBt9WVbC1E4REuJlAFtoHDT5bNYBmvdZYdIYHVVkB3brADKXsYDacoxrwpjeq9mlWsnHbSfCG_jG9T93uxqG6656ibFTdRXNa2JRX3VTKVOsWy9mclk2SRSM2KgLYhfmg2OGEOxE3-R4qO0Bg1ySeaIxqS-iT2Bg)

</details>

**Comment by barneygale** [comment]
At a depth of 400, on my system, 50% of the time is spent in `__hash__()`. This is because `_RecursiveWildcardSelector` uses a `set` to de-duplicate. We could drop the set if we handle symlinks uniformly - see #77609.

**Comment by barneygale** [comment]
I believe that the following PRs will need to land before we can properly address this issue:

- #100282
- #101398
- #102616
- #102476

**Comment by barneygale** [comment]
It occurs to me that we only need to de-duplicate results if **two** or more `**` tokens appear in the pattern. So we can improve the situation even without making symlink handling more consistent. I'll implement this soon, just need #102710 to land first I think.

**Comment by barneygale** [comment]
My plan here is:

1. Address #73435
   - PR available: #101398
2. Address #81079
   - PR available: #102710
3. Implement `glob()` using `walk()` to avoid hitting recursion errors (this issue.)
4. Address #77609
5. Address #65238
6. Implement `**/` tokens using a "walk-and-filter" strategy (this issue.)

**Comment by barneygale** [comment]
Update to my previous plan:

1. Address #81079
   - PR: #102710
2. De-duplicate paths only when necessary (this issue.)
   - PR: #104244
3. Address #90208
   - PR: #104141
4. Use `walk()` to implement recursive tokens (this issue.)
5. Address #77609
   - PR: #104176
6. Use `match()` to implement symlink-following recursive tokens with subsequent segments (this issue.)

The following can be considered future work. They will speed up recursive globbing by allowing the walk-and-match strategy to support more patterns. They don't block this issue, though.


1. Address #73435
   - PR: #101398
2. Address #65238

**Comment by barneygale** [comment]
Final PR available:

- #104512

As a result of these changes, using the repro case in the original post with 1000 nested `a/` directories, pathlib is now about **25x faster** than `glob.glob()`, rather than 150x slower! 🎉

**Comment by barneygale** [comment]
Tis fixed! Most changes made it into 3.12, but a couple are 3.13-only.
